### PR TITLE
Feature/lattice 2999 patch studies

### PR DESCRIPTION
**Pull request by solomontang**

### CODE REVIEW

**Review by geekbeast — APPROVED** [review]

**Comment by geekbeast** [comment]
So this updates a study, but ignores the organizations field in a study. This is fine, but we need to make sure javadoc for API notes that this API is not intended to update the organizations a study is associated with. 

If it is intended to handle that case let's make sure we track that work item so we don't forget it once the association API is ready (i.e it will be called from here).

**Comment by UnsungHero97** [comment]
@geekbeast ... @solomontang and I briefly talked about updating studies and decided to basically scrap this approach and use a dedicated `UpdateStudy` object instead.

**Comment by geekbeast** [comment]
What's the difference between a study object and an update study object? A bunch of optional fields?

**Comment by UnsungHero97** [comment]
Basically. The existing `Study` object has defaults which makes it tricky to work with if you want to make a request to update only a subset, similar to `MetadataUpdate`:

https://github.com/openlattice/openlattice/blob/develop/api/src/main/kotlin/com/openlattice/edm/requests/MetadataUpdate.kt

**Comment by geekbeast** [comment]
This makes sense, but the MetadataUpdate object was awful to work with on the back end especially if it had to be serialized. Consider doing it 1 by 1, but this should be pretty minor and I'm good with either approach. Assigning to you to merge if this needs to be merged.

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt [hunk 1]
```diff
@@ -136,4 +182,22 @@ class StudyService(
...
+        ps.setObject(4, study.startedAt)
+        ps.setObject(5, study.endedAt)
+        ps.setObject(6, study.lat)
+        ps.setObject(7, study.lon)
+        ps.setObject(8, study.group)
+        ps.setObject(9, study.version)
+        ps.setObject(10, study.contact)
+        ps.setString(11, mapper.writeValueAsString(study.settings))
+        ps.setObject(12, studyId)
+        ps.executeUpdate()
```
Comment (UnsungHero97) [inline src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:201]: Not every field should be set with `setObject`. For example, `title` and `description` should be `setString`. Also, I'm pretty sure we've had bugs where the index has been wrong over time when it's managed manually. Using a variable will prevent that:

```
var index = 1
ps.setString(index++, study.title)
ps.setString(index++, study.description)
...
ps.setObject(index, studyId)
```

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt [hunk 2]
```diff
@@ -116,7 +161,8 @@ class StudyService(
         ps.setObject(5, study.lon)
         ps.setObject(6, study.group)
         ps.setObject(7, study.version)
-        ps.setString(8, mapper.writeValueAsString(study.settings))
+        ps.setObject(8, study.contact)
+        ps.setString(9, mapper.writeValueAsString(study.settings))
```
Comment (UnsungHero97) [inline src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:165]: Same comment about using a variable for the index. Also, `created_at` is missing.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt (hunk 2 above)
Comment (solomontang) [reply src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:165]: createdAt is not supposed to change and is omitted from being updated.

Comment (UnsungHero97) [reply src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:165]: This is for insert.

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt [hunk 3]
```diff
@@ -50,9 +51,31 @@ class StudyService(
...
+            PostgresColumns.STUDY_VERSION,
+            PostgresColumns.CONTACT,
+            PostgresColumns.SETTINGS
+        )
+
+        private val UPDATE_STUDY_COLUMNS = UPDATE_STUDY_COLUMNS_LIST.joinToString(",") { it.name }
+
+        private val COALESCED_STUDY_COLUMNS = UPDATE_STUDY_COLUMNS_LIST
+            .joinToString(",") {
+            "coalesce(?${if(it.equals(PostgresColumns.SETTINGS)) "::jsonb" else ""}, ${it.name})"
```
[the remark this answers is not in the document]
Comment (solomontang) [reply src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:76]: `coalesce(?, <column_name>)` will yield a non null value as specified by the params, or the existing value in the column.

The if statement inside is to cast the settings string to jsonb

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt (hunk 2 above)
Comment (UnsungHero97) [reply src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:165]: I'm wrong. The postgres default takes care of `created_at`.

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt [hunk 4]
```diff
@@ -185,19 +185,24 @@ class StudyService(
...
-        ps.setString(11, mapper.writeValueAsString(study.settings))
-        ps.setObject(12, studyId)
+        var index = 1;
+        ps.setString(index++, study.title)
+        ps.setString(index++, study.description)
+        ps.setObject(index++, OffsetDateTime.now())
+        ps.setObject(index++, study.startedAt)
+        ps.setObject(index++, study.endedAt)
+        ps.setObject(index++, study.lat)
+        ps.setObject(index++, study.lon)
```
Comment (UnsungHero97) [inline src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:195]: Why not `.setDouble()`?

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (geekbeast) [reply src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:76]: Fine for this purpose, but just note this triggers a read of the row on update.

**Review by geekbeast — APPROVED** [review]

File: src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt (hunk 4 above)
Comment (solomontang) [reply src/main/kotlin/com/openlattice/chronicle/services/studies/StudyService.kt:195]: It is unhappy because the field could also be null

**Review by UnsungHero97 — APPROVED** [review]

**Pull request merged**